File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/OfflineJournalException.java
package org.hyperledger.iroha.android.offline;

/** Checked error raised by {@link OfflineJournal} operations. */
public final class OfflineJournalException extends Exception {

  private static final long serialVersionUID = 1L;

  /** Enumerates the failure categories surfaced by the journal. */
  public enum Reason {
    INVALID_KEY_LENGTH,
    INVALID_TX_ID_LENGTH,
    DUPLICATE_PENDING,
    ALREADY_COMMITTED,
    NOT_PENDING,
    IO,
    INTEGRITY_VIOLATION,
    PAYLOAD_TOO_LARGE
  }

  private final Reason reason;

  OfflineJournalException(final Reason reason, final String message) {
    super(message);
    this.reason = reason;
  }

  OfflineJournalException(final Reason reason, final String message, final Throwable cause) {
    super(message, cause);
    this.reason = reason;
  }

  public Reason reason() {
    return reason;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/OfflineJournalException.kt
package org.hyperledger.iroha.sdk.offline

/** Checked error raised by `OfflineJournal` operations. */
class OfflineJournalException : Exception {

    val reason: Reason

    internal constructor(reason: Reason, message: String) : super(message) {
        this.reason = reason
    }

    internal constructor(reason: Reason, message: String, cause: Throwable) : super(message, cause) {
        this.reason = reason
    }

    /** Enumerates the failure categories surfaced by the journal. */
    enum class Reason {
        INVALID_KEY_LENGTH,
        INVALID_TX_ID_LENGTH,
        DUPLICATE_PENDING,
        ALREADY_COMMITTED,
        NOT_PENDING,
        IO,
        INTEGRITY_VIOLATION,
        PAYLOAD_TOO_LARGE,
    }
}
